File: projects/03-patient-vitals/backend/dashboard/dynamoHelper.js
"use strict";

const { QueryCommand } = require("@aws-sdk/lib-dynamodb");

async function recentVitalWindows(chart, chartTable, vitalType, limit) {
  const resp = await chart.send(new QueryCommand({
    TableName: chartTable,
    KeyConditionExpression: "sensor_type = :st",
    ExpressionAttributeValues: { ":st": vitalType },
    ScanIndexForward: false,
    Limit: limit,
  }));
  return (resp.Items || []).reverse();
}

async function buildWardRoster(chart, chartTable, vitalTypes, patientIds) {
  const roster = new Map(patientIds.map((id) => [id, { patient_id: id, vitals: {} }]));

  for (const vitalType of vitalTypes) {
    const windows = await recentVitalWindows(chart, chartTable, vitalType, 40);
    const latestPerPatient = new Map();
    for (const windowRow of windows) latestPerPatient.set(windowRow.site_id, windowRow);

    for (const [patientId, windowRow] of latestPerPatient) {
      if (!roster.has(patientId)) roster.set(patientId, { patient_id: patientId, vitals: {} });
      roster.get(patientId).vitals[vitalType] = {
        latest: windowRow.latest,
        min: windowRow.min,
        max: windowRow.max,
        avg: windowRow.avg,
        count: windowRow.count,
        unit: windowRow.unit,
        window_end: windowRow.window_end,
        alerts: windowRow.alerts || [],
      };
    }
  }

  return Array.from(roster.values()).sort((a, b) => a.patient_id.localeCompare(b.patient_id));
}

module.exports = { recentVitalWindows, buildWardRoster };

File: projects/02-industrial-equipment/backend/dashboard/src/main/java/com/fec/industrial/dashboard/DynamoHelper.java
package com.fec.industrial.dashboard;

import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
import software.amazon.awssdk.services.dynamodb.model.QueryRequest;

import java.util.*;
import java.util.stream.Collectors;

public class DynamoHelper {

    public static Object decodeAttr(AttributeValue av) {
        if (av.s() != null) return av.s();
        if (av.n() != null) return Double.parseDouble(av.n());
        if (av.hasL()) return av.l().stream().map(DynamoHelper::decodeAttr).collect(Collectors.toList());
        return null;
    }

    public static Map<String, Object> decodeItem(Map<String, AttributeValue> item) {
        Map<String, Object> out = new LinkedHashMap<>();
        item.forEach((k, v) -> out.put(k, decodeAttr(v)));
        return out;
    }

    public static List<Map<String, Object>> recentRollups(DynamoDbClient client, String tableName,
                                                             String sensorType, int limit) {
        // scanIndexForward(false) reads newest-first for the LIMIT, then reverse restores chronological order for callers.
        QueryRequest req = QueryRequest.builder()
            .tableName(tableName)
            .keyConditionExpression("sensor_type = :st")
            .expressionAttributeValues(Map.of(":st", AttributeValue.fromS(sensorType)))
            .scanIndexForward(false)
            .limit(limit)
            .build();
        List<Map<String, Object>> items = client.query(req).items().stream()
            .map(DynamoHelper::decodeItem)
            .collect(Collectors.toList());
        Collections.reverse(items);
        return items;
    }

    public static Map<String, Object> assetSummary(DynamoDbClient client, String tableName, String[] sensorTypes) {
        List<Object> sensors = new ArrayList<>();
        for (String sensorType : sensorTypes) {
            // Take the last 20 windows and keep only each site's latest entry; oldest-first order lets later windows overwrite.
            List<Map<String, Object>> recent = recentRollups(client, tableName, sensorType, 20);
            Map<String, Map<String, Object>> bySite = new LinkedHashMap<>();
            for (var item : recent) bySite.put((String) item.get("site_id"), item);

            List<Object> sites = new ArrayList<>();
            List<String> siteIds = new ArrayList<>(bySite.keySet());
            Collections.sort(siteIds);
            for (String siteId : siteIds) {
                Map<String, Object> item = bySite.get(siteId);
                Map<String, Object> site = new LinkedHashMap<>();
                site.put("site_id", siteId);
                site.put("latest", item.get("latest"));
                site.put("unit", item.get("unit"));
                site.put("min", item.get("min"));
                site.put("max", item.get("max"));
                site.put("count", item.get("count"));
                site.put("window_end", item.get("window_end"));
                site.put("alerts", item.get("alerts"));
                sites.add(site);
            }
            Map<String, Object> sensor = new LinkedHashMap<>();
            sensor.put("sensor_type", sensorType);
            sensor.put("sites", sites);
            sensors.add(sensor);
        }
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("sensors", sensors);
        return result;
    }
}
